compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/CharHelpers.skiko.kt

```diff
@@ -150,3 +152,107 @@ internal fun String.codePointBefore(index: Int): CodePoint {
     }
     return low.code
 }
+
+/**
+ * Finds the offset of the next non-whitespace symbols subsequence (word) in the given text
+ * starting from the specified caret offset.
+ *
+ * @param offset The offset where to start looking for the next word.
+ * @param currentText The current text in which to search for the next word.
+ * @return The offset of the next non-whitespace symbols subsequence (word), or the end of the string
+ *         if no such word is found.
+ */
+@ExperimentalTextApi
+fun findNextNonWhitespaceSymbolsSubsequenceStartOffset(
+    offset: Int,
+    currentText: String
+): Int {
+    /* Assume that next non whitespaces symbols subsequence (word) is when current char is whitespace and next character is not.
+     * Emoji (compound incl.) should be treated as a new word.
+     */
+    val charIterator =
+        BreakIterator.makeCharacterInstance() // wordInstance doesn't consider symbols sequence as word
+    charIterator.setText(currentText)
+
+    var currentOffset: Int
+    var nextOffset = charIterator.next()
+    while (nextOffset < offset) {
+        nextOffset = charIterator.next()
+    }
+    currentOffset = nextOffset
+
+    while (nextOffset != BreakIterator.DONE) {
+        nextOffset = charIterator.next()
+        if (currentText.codePointAt(currentOffset).isWhitespace() && !currentText.codePointAt(
+                nextOffset
+            ).isWhitespace()
+        ) {
+            return currentOffset
+        } else {
+            currentOffset = nextOffset
+        }
+    }
+    return currentOffset
+}
+
+/**
+ * Determines whether the character at the specified offset in the string is a whitespace Unicode character.
+ *
+ * @param offset The index of the character to check.
+ * @return `true` if the character at the specified offset is a whitespace character, `false` otherwise.
+ */
+@ExperimentalTextApi
+fun String.isWhitespace(offset: Int): Boolean {
```

Comment:
We shouldn't expose `String` extensions as public API (despite of any annotations).
Maybe we just can move more high-level function (like `determineCursorDesiredOffset`) to ui module (not to this particular file) to avoid exposing functions that require offsets/codepoints?

Also, I'd avoid duplicating it as `String` extension at all, because it's not clear that it will check more than one character.

Reply:
Can we move these functions to `foundation` and mark internal?

Reply:
then we will need to public/copy all low level unicode (code point) functions since it's still not in std lib

Reply:
How much code we should copy?

The API design approach in Compose between layers - if it needs only for other layer - we either design a real public API for that (and maintain backward compatibility), or copy all implementations.

The copying is simplier in our case.

Reply:
Everything related to CodePoints (130 lines of code).

Reply:
Moved only needed code related to CodePoints to StringHelpers.skiko.kt in foundation/text.